### ISSUE TITLE
type arguments, classes, and coercions

### ISSUE THREAD

**Issue by mppf** [chapel-lang/chapel#13541]
PR #13447 implements - as a starting point - certain coercion rules for type arguments considering nilability and management type. However these rules need more discussion and refinement.

Relates to #7779 and #8083.

First of all, after PR #8083, the following program does not compile and run:
``` chapel
proc foo(type t:int) { }
foo(int(8));
```

because the coercion from `int(8)` to `int` is not allowed for a type argument.

However PR #8083 made an exception for coercions to parent class types, so that the following compiles and runs:
``` chapel
class Parent { }
class Child : Parent { }
proc bar(type t:borrowed Parent) { }
bar(borrowed Child);
```

(here borrowed Child coerces to borrowed Parent as a type argument). The justification for this was that it would behave similarly to a where clause with the type constraint (which at that time would have been written `proc bar(type t) where t: borrowed Parent`).

However, since then, there have been 2 things that require us to revisit this question:
 1.  `:` in where clauses has been replaced by `isSubtype` and `isSubtype` *does* consider coercions (see #9273 and #10060)
 2. There are now more variants of class types for which one might want to allow coercions:
  * coercions on management type, e.g. owned -> borrowed
  * coercions on nilability, e.g. not-nilable C -> nilable C?

Due to (2), PR #13447 allows coercions among variants of class types in type arguments.
For example, the following compiles and runs:
``` chapel
class Parent { }
class Child : Parent { }
proc baz(type t:Parent?) { }
baz(borrowed Child);
```

and even this compiles and runs:
``` chapel
class Parent { }
class Child : Parent { }
proc boop(type t:borrowed Parent?) { }
boop(Child);
```

Should we try to:
1. never apply coercions to type arguments
2. apply coercions to class types only
3. always apply coercions to type arguments

If we choose to continue with (2), there are additional questions about which types of coercions on class arguments should be allowed. Nilability? Owned to borrowed? Owned to parent class type owned? Generic management to specific management?

**Comment by bradcray** [comment]
In reading this, my initial reaction was ("since Michael is asking about this") that option 2 seemed like an unfortunate middle-ground and that perhaps we should go with option 3.  But that came in part from a sense that I wasn't sure type constraints on value types are really all that useful or all that likely to be used, so maybe it "just didn't matter."

But then in talking through where they are used today (where the two main answers are "when defining casts" and "when expressing constraints on the `this` argument to type methods), I started to get cold feet about having value types coerce.

An example I was musing about was "What if I were to implement a "numMantissaBits()" method for Chapel's real types such that I could write things like `real(64).numMantissaBits()` or `real(32).numMantissaBits()`.  Intuitively, I don't think I'd want those routines to permit me to say `int.numMantissaBits()`, but permitting `type int` to coerce to `type real` seems like it would cause that to happen.  So the implication is that rather than writing a body like:

```chapel
if this.numBits() == 64 then return 52;
if this.numBits() == 32 then return 23;
```

I'd also have to code defensively to make sure that this wasn't called on an `int` (or anything else that could coerce to real), either by overloading the method for int / other coerce-to-real types; or by defending against it in the body of the method.

Then I started thinking about how I'd actually write this function and started to get more confused.  For example, if I wrote it like this:

```chapel
proc real.numBitsInMantissa() { ... }
```

it seems like I'm really saying "define this for 64-bit reals" in which case I either wouldn't be able to call it on `type real(32)`, or `type real(32)` would coerce to `type real(64)` and I would get the wrong answer (since the body of the function would think we were in the 64-bit case).

The other way I could write it might be like this:

```chapel
proc (real(?)).numBitsInMantissa() { ... }
```

which seems like it solves the `real(32)` vs. `real(64)` issue, but seems less like it should support calls from `int` type.

Another similar example was thinking about casts.  If I created a record type `R` and created an `R-to-real` cast, I don't think I would want `myR: int` to call the cast-to-real routine and return a real to me.  And as with the previous example, I don't think I'd want to have to defensively worry about preventing this case by creating cast-to-int and cast-to-anything-that-can-coerce-to-type-R overloads to protect myself.

So this made me think we shouldn't support coercions on type arguments, for value types.

At this point in the conversation, Michael asked whether my examples above would also apply in a parent/child class relationship, e.g., where `real` was replaced by `Parent` and `int` was replaced by `child`.  I didn't have an answer that and considered it a "What do other OOP types do in this regard?" type of question (which he said he'd look into).  I also have a vague memory of @vasslitvinov  asking either this same question or a related one recently, but can't quite remember the context or outcome.

So that suggests that depending on the answer to "What does Smalltalk / Java / whatever OOP-with-class-hierarchies-and-type methods do" I'd either fall into option 1 or 2 with my current mindset.

**Comment by mppf** [comment]
Java:

It does allow the equivalent of type methods to be called on the child type (implying a kind of coercion).

``` java
// Test.java
public class Test {
  public static void main(String[] args) {
    TestChild c = new TestChild();
    c.parentTypeMethod();

    TestChild.parentTypeMethod();
  }
}
```

``` java
// TestParent.java
public class TestParent {
  public static void parentTypeMethod() {
    System.out.println("in parentTypeMethod()");
  }
}
```

``` java
/// TestChild.java
public class TestChild extends TestParent {
}
```

**Comment by vasslitvinov** [comment]
Here is one route that is self-consistent:

* Treat the formal decl `type t1: t2` as another way of saying `where isSubtype(t1, t2)` .
* Treat `isSubtype(t1,t2)` says whether a variable of the type `t1` can be passed to an `in`-intent formal of the type t2.
  - or, equivalently,  const-in or const-ref intent

This route takes us to Option 3.

This made me think about the type of the instantiated generic formal whose declared type is `t2` and whose actual is of the type `t1`. Should the instantiated formal's type be an instantiation of `t2` (if applicable) or should it be `t1` ?

#13154 is where I asked the question of inheriting a type method in child classes. The most recent implementation changes are in #13158. 

Inheriting a type method by a child class is different from inheriting a type method by `int` from `real`.  In the former case, the child class has supposedly everything that the parent class has. In the latter case, `int` lacks some things that `real` has, such as mantissa. Ditto inheriting by a record that is coercible to a number.

So maybe this calls for providing a separate predicate that allows instantiations while disallowing coercions and subtyping, unlike `isSubtype`.

**Comment by mppf** [comment]
@vasslitvinov - thanks for the link to #13154. That leaves the answer to "what do other languages do" as:

 * C++, Swift, Java allow type methods defined on a parent to be invoked on a child class.

Back to @bradcray's comment

> So that suggests that depending on the answer to "What does Smalltalk / Java / whatever OOP-with-class-hierarchies-and-type methods do" I'd either fall into option 1 or 2 with my current mindset.

I think this means that "never apply coercions" is not the right move for us here. That suggests that @bradcray would prefer option 2 - apply coercions to class types only.

This leaves me wondering if this is just an exception for classes or if there is something more fundamental going on here.

While the Chapel compiler treats subtyping and coercions as the same, I can think of one way to define the terms that has a concrete difference. Neither a subtype nor a coercion can pass to a `ref` formal (because the body of the function could for example replace an allocated Child with an allocated Parent, causing code after the call to not be type-safe).

We could define the term *subtype* as something that can pass to a `const ref` formal while the `const ref` formal can still refer to part of original value. Meaning that a function accepting the value by `const ref` could observe modifications to that value during its execution (e.g. supposing it was passed a global and then called a function that changed the global). This works when the parent type is physically contained in the original value in memory. This is how we implement classes.  See also #8489 which discusses passing e.g. a `Child` value to a `const ref Parent` argument, which we don't allow today, but that has more to do with the C backend than with Chapel's type system.

In contrast, a coercion from `int` to `real` does not have this relationship. Neither the `int` nor the `real` is stored in memory within the other. As a result, a coercion cannot pass to a `const ref` formal if we want it to refer to the original value.

I am uncertain of the proper terms to use for these properties, but if we decided to hang our hat on this difference, perhaps we would split `isSubtype` into `isCoercible` and `isSubtype`. (I would prefer we leave `isSubtype` containing can-instantiate for when the RHS is generic).

Supposing we did that and stuck with option 2, what would that mean for the related questions:

> there are additional questions about which types of coercions on class arguments should be allowed. Nilability? Owned to borrowed? Owned to parent class type owned? Generic management to specific management?

### Nilability

In this context, we are talking specifically about coercions related to nilability. Those coercions are from a non-nilable actual to a nilable formal. E.g.

``` chapel
proc f(const ref arg: borrowed MyClass?) {
}
var x: borrowed MyClass = ...;
f(x); // applies coercion from MyClass to MyClass?
```

Since all nilability for classes is removed before the aliasing rules apply, we could allow a `const ref` to point to something declared with the other nilability. However, for things like `int?` which we imagine are one day implemented as an `Optional`, the `Optional(int)` would contain the memory necessary to store the `int`, so it'd be possible to get a `const ref` to the `int` from an `int?`. However, one can't find an `int?` in the same memory as an `int`, which is the direction we want to go here.

As a result I would consider `something` -> `something?` to be a coercion and not subtyping.

### Owned to borrowed

Since `owned` contains the pointer memory, a `const ref` to that could be returned in borrowing. So `owned MyClass` -> `borrowed MyClass` could be considered subtyping

### owned Child to owned Parent

Here, the `owned` record would use exactly the same memory representation when referring to a particular value whether that record is typed `Child` or `Parent`. So this is subtyping (but we can't actually implement it that way with the C backend, and would need to adjust LLVM TBAA to actually allow it, see #8489).

### Generic management to specific management

Since all management flavors contain the pointer, we can consider `anymanaged MyClass` -> `borrowed MyClass` to be subtyping as well.

**Comment by vasslitvinov** [comment]
I support Michael's reasoning based on ability to pass by `const ref`, and the resulting conclusions. Sounds like class inheritance and ownership changes are the only ways to get `isSubtype`.

`isInstanceof` - maybe it will be useful, too?  Then,
```
isProperSubtype(a,b) = isSubtype(a,b) && ! isInstanceof(a,b)
```

`isCastable` - how about this? It's when one can be converted to the other via an explicit cast.

**Comment by vasslitvinov** [comment]
Back to the original topic, I suggest that we tackle the "allow coercions?" question **separately** for the two cases:

* for a type-intent formal  proc foo(type t: T)
* for the receiver of a type method  proc MyClass.foo()

For a type method, we might want to split it into two cases:

* `proc MyClass.foo()` -- it is a primary method or a secondary method declared with just the class name.
* `proc (owned MyClass?).foo()` -- it is a secondary method declared with some specifics about the class: memory management, nilability, generic instantiation if applicable.

For a type-intent formal, `:` visually calls for allowing coercions. So we could choose Option 3 so it works like it does for a generic argument:

```chpl
proc p1(arg: T)     // allows coercions,
                    // except weird cases with records coercible to ints
proc p2(type t: T)  // may as well allow coercions
                    // (except those weird cases, too?)
```

For a type method, I do not see compelling arguments to allow non-class-subtyping coercion. So we may disallow it for now until there is a use case.

Brad's `real.numMantissaBits()` example argues in favor of disallowing. If it is defined on `real` and not on `real(32)`, the users would silently get an incorrect answer when invoked on `real(32)`.

Can we make the same argument about a child class vs. a parent class? For example:

```chpl
proc type Parent.className() return "Parent";
writeln(Child.className()); // prints "Parent", perhaps unintentionally
```

OTOH Swift et al. do not consider this a show-stopper for allowing coercions. So maybe the `real.numMantissaBits()` example is not a show-stopper, either?

**Comment by mppf** [comment]
In "A behavioral notion of subtyping", Liskov and Wing argue that a notion of subtyping needs to consider the possibility that there are multiple references to the same object, and these references can have different static types. e.g.
``` chapel
var myChild = new owned myChild();
var x = myChild.borrow(): borrowed Parent;
var y = myChild.borrow(): borrowed Child;
x.parentMethod();
y.childMethod(); // the underlying object needs to still make sense
x.otherParentMethod(); // and here as well
```

This gets me thinking that the `const ref` idea above is right, because one can have aliasing even if there is not the potential for mutation through the aliases.

References to the same memory with different types can occur if and only if one type is a subtype of the other. This might not be a complete definition of *subtype* by itself, but it is a stronger property than coercibility provides and there seems to be some precedent for this sort of thinking in the literature.

**Comment by mppf** [comment]
AFAIK there is enough agreement to proceed with https://github.com/chapel-lang/chapel/issues/13541#issuecomment-517693254 which is a variant of Option 2 in the original description.

**Comment by vasslitvinov** [comment]
> there is enough agreement to proceed with #13541 (comment) which is a variant of Option 2 in the original description

I see the agreement being about the distinction between subtyping and coercion. How it applies to type arguments is a separate concern.

It seems like Option 2 is motivated primarily by "how other languages do it". This leaves me wanting a stronger motivation. However I am OK to proceed with it.

**Comment by bradcray** [comment]
> It seems like Option 2 is motivated primarily by "how other languages do it". This leaves me wanting a stronger motivation. However I am OK to proceed with it.

Do you think other languages made the wrong choice in this regard?  Do you have a case where you think option 2 causes problems?

**Comment by vasslitvinov** [comment]
> Do you think other languages made the wrong choice in this regard?

I don't know one way or the other. I am just not seeing why they made their choice.

**Comment by bradcray** [comment]
In that case, a secondary reason for option 2 is that it's how Chapel is implemented today which has the advantage of not breaking existing code or requiring it to be rewritten while also following established trends.

**Comment by mppf** [comment]
@vasslitvinov - Regarding type methods, I think that @bradcray had a pretty good user-facing / use case based rationale in https://github.com/chapel-lang/chapel/issues/13541#issuecomment-517411449 for why Option 3 (always apply coercions to type arguments) would be surprising.

So, that leaves us choosing between Option 1 (never apply coercions to type arguments and Option 2 (do it only for certain class patterns).

If we have the viewpoint that inheritance causes subtypes to inherit behavior available on the parent type, then calling a plain (non-type) method is one such behavior, right? Arguably, calling a method defined only on the parent type requires a "coercion" (in terms of the compiler's implememantion) but it is something we can view as "subtyping" according to the rationale in https://github.com/chapel-lang/chapel/issues/13541#issuecomment-517693254 .

Now, this reasoning does generalize to type methods because OO languages normally allow type methods to be called on instances. I.e., I can call`myInstance.someMethod()` when `someMethod` is a type method (aka static method in Java, e.g.). Now, what would happen if `someMethod` was a type method that was defined only on the parent type? Our compiler would view it as a coercion to go from the child type to the parent type (but again, I think we should call it "subtyping" here). If we do not allow this type of "subtyping", it would fail to resolve. That would be surprising.

So, in other words, I think that Option 2 comes from:
 * a desire to avoid the confusing situation Brad brought up with Option 3
 * while respecting the idea that inheritance makes Parent methods available on a Child type
 * and respecting the idea that type methods are callable on instances
    * and so behave similarly to instance methods w.r.t inheritance; in other words the Parent method will be called if the Child has defined none. 

Another way to put that is - it would be a poor notion of "subtype" if there was a behavior available on the parent type (call a type method) that was not available on the child type.

**Comment by bradcray** [comment]
As a potential aside:

> respecting the idea that type methods are callable on instances

Note that Chapel doesn't support calling type methods on instances (e.g., [TIO](https://tio.run/##NYpBCoAgFAX3/xQPV7rpAtHKVccQMQpMRSWR8OxmQbthZvSugrG9a6tSgsRNQIheI9dgsHnPxeeAEo9srONsdZDTKEzMIzRqRJeKOKvEAmcKJB/lW14Y/sfeHw)).  I tend to like this from a purity / simplicity / avoid potential confusion perspective, but maybe I just don't understand the traditional rationale for supporting it in typical languages.

In saying this, my intention isn't to undermine support for option 2 which still seems most preferable to me.

**Comment by vasslitvinov** [comment]
Option 2 makes a good sense because it allows inheriting a type method. If we have a good motivation for "inheriting" where coercion is allowed, we can revisit Option 3 at that time.

**Comment by mppf** [comment]
We are thinking of changing isSubtype into isSubtype and isCoercible. But we also have `<=` as an alias for `isSubtype`. Should this check the new `isSubtype` or the new `isCoercible` ? (Or should we remove these type comparison operators?)

Note, I am expecting isCoercible to include isSubtype (i.e. if isSubtype returns true, so does isCoercible).

**Comment by bradcray** [comment]
My intuition is that `<=` should mean `isSubtype`, but that's just a gut reaction.  Tagging @benharsh who I think implemented the feature originally.

**Comment by vasslitvinov** [comment]
One way to think about it is to make `<=` mean the test that is used most frequently. For example, I would expect `<=` to be the thing I will want to use in a where-clause most of the time.

**Comment by mppf** [comment]
Note that if `<=` includes coercions, it will behave the same as in the last release.

@vasslitvinov 

> One way to think about it is to make <= mean the test that is used most frequently. 

Right, but do you expect you will most frequently want to know if something is coercible or if it is a stricter subtype? I don't think we have much existing code that cares about the difference. (At least, testing passes with the behavior switched).

Usually in code I have written, I have used `isSubtype` in preference to `<=` in where clauses because I think it is clearer to a reader.

We discussed `<=` on types allowing coercions in #10060. At the moment I am leaning towards making `<=` map to `isSubtype` since `<=` meaning "is subtype" has more precedent AFAIK than it meaning "is coercible". My previous viewpoint was that "is subtype" includes coercion as a kind of subtyping, but with the isSubtype/isCoercible split we are no longer doing that. Even so, this choice feels a bit arbitrary to me and I'd appreciate other viewpoints.

Another way to put the question is - should `int(8) <= int(16)` return `true`? On master it does, but if we make `<=` mean the new isSubtype, it will return `false`.

Any thoughts here @benharsh ?

**Comment by benharsh** [comment]
I think it would be good to see the potential impact of this change before deciding on a particular direction.

I have a slight preference for ``<=`` mapping to ``isSubtype``.

**Comment by mppf** [comment]
@benharsh - PR #13813 updates `<=` to map to the stricter `isSubtype` and the only test or module code I needed to update was in tests of `<=` on types itself.
